src/expr/src/vector_op/bitwise_op.rs

```diff
@@ -0,0 +1,109 @@
+// Copyright 2022 Singularity Data
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+use std::any::type_name;
+use std::convert::TryInto;
+use std::fmt::Debug;
+use std::ops::{BitAnd, BitOr, BitXor, Not};
+
+use num_traits::{CheckedShl, CheckedShr};
+use risingwave_common::error::ErrorCode::{InternalError, NumericValueOutOfRange};
+use risingwave_common::error::{Result, RwError};
+
+use crate::vector_op::arithmetic_op::general_atm;
+
+#[inline(always)]
+pub fn general_shl<T1, T2, T3>(l: T1, r: T2) -> Result<T3>
+where
+    T1: TryInto<T3> + Debug,
+    T2: TryInto<u32> + Debug,
+    T3: CheckedShl,
+{
+    general_shift(l, r, |a, b| match a.checked_shl(b) {
+        Some(c) => Ok(c),
+        None => Err(RwError::from(NumericValueOutOfRange)),
+    })
+}
+
+#[inline(always)]
+pub fn general_shr<T1, T2, T3>(l: T1, r: T2) -> Result<T3>
+where
+    T1: TryInto<T3> + Debug,
+    T2: TryInto<u32> + Debug,
+    T3: CheckedShr,
+{
+    general_shift(l, r, |a, b| match a.checked_shr(b) {
+        Some(c) => Ok(c),
+        None => Err(RwError::from(NumericValueOutOfRange)),
+    })
+}
+
+#[inline(always)]
+pub fn general_shift<T1, T2, T3, F>(l: T1, r: T2, atm: F) -> Result<T3>
+where
+    T1: TryInto<T3> + Debug,
+    T2: TryInto<u32> + Debug,
+    F: FnOnce(T3, u32) -> Result<T3>,
+{
+    // TODO: We need to improve the error message
+    let l: T3 = l.try_into().map_err(|_| {
+        RwError::from(InternalError(format!(
+            "Can't convert {} to {}",
+            type_name::<T1>(),
+            type_name::<T3>()
+        )))
+    })?;
+    let r: u32 = r.try_into().map_err(|_| {
+        RwError::from(InternalError(format!(
+            "Can't convert {} to {}",
+            type_name::<T2>(),
+            type_name::<u32>()
+        )))
+    })?;
+    atm(l, r)
+}
```

Comment:
As mentioned above, the left side does not need cast, and the right side can be int16 or int32 - `Into<i32>`.
There seems to be no overflow error, and negative right side have sign bit discarded:
```
test=# select 1::smallint << 14;
 ?column? 
----------
    16384
(1 row)

test=# select 1::smallint << 15;  -- overflow to negative
 ?column? 
----------
   -32768
(1 row)

test=# select 1::smallint << 16;  -- overflow to truncated
 ?column? 
----------
        0
(1 row)

test=# select 1::smallint << 2147483647;  -- overflow to truncated, same as above
 ?column? 
----------
        0
(1 row)

test=# select 1::smallint << -2147483648;  -- same as `<< 0`
 ?column? 
----------
        1
(1 row)

test=# select 1::smallint << -2147483647;  -- same as `<< 1`
 ?column? 
----------
        2
(1 row)
```

Reply:
overflow to truncated is implemented. However, 
 test=# select 1::smallint << -2147483648;  -- same as `<< 0`
test=# select 1::smallint << -2147483647;  -- same as `<< 1`

So the converting -2147483648 to 0 and -2147483647 to 1  is by chopping off the sign bit?

I can,t do `Into<i32>` because ` .checked_shl() ` only allow u32. Besides, rust seems to can,t convert i32 to u32. if the number is negative.

I try `std::ops::{Shl,Shr}` too and same problem.

Any idea for solving this? Except I have to do the algo that converts i32 to u32.

Reply:
I think we need to impl this special i32 to u32 conversion here.

Reply:
It is undefined in C or C++, but I thought PostgreSQL is doing sanitization before using the C operator, just like they did for `+` ([ref](https://github.com/postgres/postgres/blob/73508475d69e90f98ebd9b7e1a5933a26a49c5e9/src/backend/utils/adt/int.c#L908-L920)).
However, it is only doing that for `+` but not for `<<` ([ref](https://github.com/postgres/postgres/blob/73508475d69e90f98ebd9b7e1a5933a26a49c5e9/src/backend/utils/adt/int.c#L1484-L1491)). My bad ...

Based on the latest finding, shall we ...
* Revert to the behavior of returning overflow error, as it is also undefined behavior
* Leave some comments about this conscious decision to diverge from PostgreSQL - we are not forced to do it differently simply because we are using a different language, but because PostgreSQL behavior is harmful.